Source/WebCore/platform/graphics/gstreamer/GStreamerVideoSinkCommon.cpp

```diff
@@ -44,31 +44,46 @@ class WebKitVideoSinkProbe {
     }
 
+    void handleFlushEvent([[maybe_unused]] GstPad* pad, GstPadProbeInfo* info)
+    {
+        // We need to operate from the main thread, this is a requirement for usage of
+        // AbortableTaskQueue start/finishAborting.
+        ASSERT(isMainThread());
+
+        if (GST_EVENT_TYPE(GST_PAD_PROBE_INFO_EVENT(info)) == GST_EVENT_FLUSH_START) {
+            if (!m_isFlushing) {
+                GST_DEBUG_OBJECT(pad, "FLUSH_START received, aborting all pending tasks in the player sinkTaskQueue.");
+                m_isFlushing = true;
+                m_player->sinkTaskQueue().startAborting();
+#if USE(GSTREAMER_GL)
+                    GST_DEBUG_OBJECT(pad, "Flushing current buffer in response to %" GST_PTR_FORMAT, info->data);
+                    m_player->flushCurrentBuffer();
+#endif
+            } else
+                GST_DEBUG_OBJECT(pad, "Received FLUSH_START while already flushing, ignoring.");
+            return;
+        }
+
+        RELEASE_ASSERT(GST_EVENT_TYPE(GST_PAD_PROBE_INFO_EVENT(info)) == GST_EVENT_FLUSH_STOP);
+        if (m_isFlushing) {
+            GST_DEBUG_OBJECT(pad, "FLUSH_STOP received, allowing operation in the player sinkTaskQueue again.");
+            m_isFlushing = false;
+            m_player->sinkTaskQueue().finishAborting();
+            return;
+        }
+
+        GST_DEBUG_OBJECT(pad, "Received FLUSH_STOP without a FLUSH_START, ignoring.");
+    }
+
     static GstPadProbeReturn doProbe([[maybe_unused]] GstPad* pad, GstPadProbeInfo* info, gpointer userData)
     {
         auto* self = static_cast<WebKitVideoSinkProbe*>(userData);
         auto* player = self->m_player;
 
         if (info->type & GST_PAD_PROBE_TYPE_EVENT_FLUSH) {
-            if (GST_EVENT_TYPE(GST_PAD_PROBE_INFO_EVENT(info)) == GST_EVENT_FLUSH_START) {
-                if (!self->m_isFlushing) {
-                    GST_DEBUG_OBJECT(pad, "FLUSH_START received, aborting all pending tasks in the player sinkTaskQueue.");
-                    self->m_isFlushing = true;
-                    player->sinkTaskQueue().startAborting();
-#if USE(GSTREAMER_GL)
-                    GST_DEBUG_OBJECT(pad, "Flushing current buffer in response to %" GST_PTR_FORMAT, info->data);
-                    player->flushCurrentBuffer();
-#endif
-                } else
-                    GST_DEBUG_OBJECT(pad, "Received FLUSH_START while already flushing, ignoring.");
-            } else if (GST_EVENT_TYPE(GST_PAD_PROBE_INFO_EVENT(info)) == GST_EVENT_FLUSH_STOP) {
-                if (self->m_isFlushing) {
-                    GST_DEBUG_OBJECT(pad, "FLUSH_STOP received, allowing operation in the player sinkTaskQueue again.");
-                    self->m_isFlushing = false;
-                    player->sinkTaskQueue().finishAborting();
-                } else
-                    GST_DEBUG_OBJECT(pad, "Received FLUSH_STOP without a FLUSH_START, ignoring.");
-            }
+            callOnMainThreadAndWait([&] {
```

Comment:
Let's add a comment so that the next person that comes here knows why this is needed. For example:

```
// Usually flushes propagate in the main thread as a synchronous consequence of a seek.
// However, this doesn't have to be the case:
//
// As a notable example, when matroskademux receives a seek before it has parsed the
// entire file header, it stores the event and returns without flushing anything.
// Later, the streaming thread finishes parsing the file header and handles the stored
// seek event from that same thread. This sends a flush from the streaming thread.
```